frame/lending/src/lib.rs

```diff
@@ -1223,183 +1356,415 @@ pub mod pallet {
+				collateral_balance_after_withdrawal_value,
+				borrow_balance_value,
+				market
+					.collateral_factor
+					.try_into_validated()
+					.map_err(|_| Error::<T>::Overflow)?, // TODO: Use a proper error mesage
+				market.under_collateralized_warn_percent,
+			);
+
+			ensure!(
+				!borrower_after_withdrawal.should_liquidate()?,
+				Error::<T>::WouldGoUnderCollateralized
+			);
+
+			let market_account = Self::account_id(market_id);
+			// ensure!(
+			// 	<T as Config>::MultiCurrency::can_deposit(market.collateral_asset, account, amount)
+			// 		== DepositConsequence::Success,
+			// 	Error::<T>::TransferFailed
+			// );
+			// ensure!(
+			// 	<T as Config>::MultiCurrency::can_withdraw(
+			// 		market.collateral_asset,
+			// 		&market_account,
+			// 		amount
+			// 	)
+			// 	.into_result()
+			// 	.is_ok(),
+			// 	Error::<T>::TransferFailed
+			// );
+
+			AccountCollateral::<T>::try_mutate(market_id, account, |collateral_balance| {
+				let new_collateral_balance =
+					// REVIEW: Should we default if there's no collateral? Or should an error (something like "NoCollateralToWithdraw") be returned instead?
+					collateral_balance.unwrap_or_default().safe_sub(&amount)?;
 
-			markets
+				collateral_balance.replace(new_collateral_balance);
+
+				Result::<(), DispatchError>::Ok(())
+			})?;
+			<T as Config>::MultiCurrency::transfer(
+				market.collateral_asset,
+				&market_account,
+				account,
+				amount,
+				true,
+			)
+			.expect("impossible; qed;");
+			Ok(())
 		}
 
-		fn get_all_markets() -> Vec<(Self::MarketId, MarketConfiguration<T>)> {
-			Markets::<T>::iter().map(|(index, config)| (index, config)).collect()
+		fn get_markets_for_borrow(borrow: Self::VaultId) -> Vec<Self::MarketId> {
+			// // allow to be slow until it becomes write transaction (not the case now and then)
+			// let mut markets = vec![];
+			// for (index, market) in Markets::<T>::iter() {
+			// 	if market.borrow == borrow {
+			// 		markets.push(index);
+			// 	}
+			// }
+
+			// markets
+
+			Markets::<T>::iter()
+				.filter_map(|(index, market)| market.borrow_asset_vault.eq(&borrow).then(|| index))
+				.collect()
 		}
 
 		fn borrow(
 			market_id: &Self::MarketId,
-			debt_owner: &Self::AccountId,
+			borrowing_account: &Self::AccountId,
 			amount_to_borrow: BorrowAmountOf<Self>,
 		) -> Result<(), DispatchError> {
 			let market = Self::get_market(market_id)?;
-			let borrow_asset = T::Vault::asset_id(&market.borrow)?;
-			let market_account = Self::account_id(market_id);
+			let MarketAssets { borrow_asset, debt_asset: debt_asset_id } =
+				Self::get_assets_for_market(market_id)?;
 
-			Self::can_borrow(market_id, debt_owner, amount_to_borrow, market, &market_account)?;
+			let market_account = Self::account_id(market_id);
 
-			let debt_asset_id = DebtMarkets::<T>::get(market_id);
-			let new_account_interest_index = Self::updated_account_interest_index(
+			// REVIEW: Remove? (remove can_* checks)
```

Comment:
there is separate task to make all [transactional] and remove checks, mark here it as todo